File: mobile/library/swift/src/grpc/GRPCClient.swift
import Foundation

/// Client that supports sending and receiving gRPC traffic.
@objcMembers
public final class GRPCClient: NSObject {
  private let streamClient: StreamClient

  /// Create a new gRPC client instance.
  ///
  /// - parameter streamClient: The stream client to use for gRPC streams.
  public init(streamClient: StreamClient) {
    self.streamClient = streamClient
  }

  /// Create a new gRPC stream prototype which can be used to start streams.
  ///
  /// - returns: The new gRPC stream prototype.
  public func newGRPCStreamPrototype() -> GRPCStreamPrototype {
    let prototype = self.streamClient.newStreamPrototype()
    return GRPCStreamPrototype(underlyingStream: prototype)
  }
}

File: mobile/library/kotlin/io/envoyproxy/envoymobile/grpc/GRPCClient.kt
package io.envoyproxy.envoymobile

// 1 byte for the compression flag, 4 bytes for the message length (int)
internal const val GRPC_PREFIX_LENGTH = 5

/**
 * Client that supports sending and receiving gRPC traffic.
 *
 * @param streamClient The stream client to use for gRPC streams.
 */
class GRPCClient(private val streamClient: StreamClient) {
  /**
   * Create a new gRPC stream prototype which can be used to start streams.
   *
   * @return The new gRPC stream prototype.
   */
  fun newGRPCStreamPrototype() = GRPCStreamPrototype(streamClient.newStreamPrototype())
}
